[16.0][FIX] users_ldap_groups: safe LDAP decode

oh2fih opened the pull request:
The group mapping in query mode fails if LDAP returns binary data in any of the fields. This adds a function that handles such situation by base64 encoding it.

oh2fih commented:
This same bug affects all other versions, too.

I'm about to give up writing the test coverage for this change because codecov keeps complaining about everything I try. (Especially I didn't understand why it demands test coverage for the tests, too.) My idea is to add binary data to the LDAP response, e.g., a minimal GIF image as the LDAP attribute `thumbnailPhoto`, as this is a common case where LDAP returns binary data.

```
{
    "dc=users_ldap_groups,dc=example,dc=com": {
        "cn": [b"User Name"],
        "name": [b"hello", b"hello2"],
        "thumbnailPhoto": [
            b"GIF89a\x01\x00\x01\x00\x00\xff\x00,"
            b"\x00\x00\x00\x00\x01\x00\x01\x00\x00\x02\x00;"
        ],
    }
}

oh2fih commented:
Tested with Odoo 16.0.20240102 on Ubuntu 22.04.

oh2fih commented:
This patch now has 100% test coverage after adding a fully functioning test where LDAP returns binary data. I'm leaning towards the conclusion that the `codecov/project` pipeline is simply broken...

oh2fih commented:
The patches have now been rearranged into coherent commits that introduce one fix at a time and adds the required test coverage in parallel, so this pull request should be easy to review.

Both the patches have been tested with Odoo 16.0.20240328 on Ubuntu 22.04. All unit tests are also passing with 100% test coverage (`100.00% of diff hit (target 91.95%)`). The pull request patches a vulnerability involving [Unsafe Public Methods](https://www.odoo.com/documentation/17.0/developer/reference/backend/security.html#unsafe-public-methods) & fixes a bug that renders the entire module useless whenever LDAP returns binary data in any of the fields.

I can see that `users_ldap_groups` is missing a named maintainer, but not fixing this kind of bugs and vulnerabilities for months surely seems like this @OCA is dead. Could, e.g., @joao-p-marques take a look at this as `users_ldap_mail` & `users_ldap_populate` are pretty close to this. Thank you!

rvalyi reviewed (approved):
I'm no ldap specialist at all, but I understand the vulnerability and your fix seems both reasonable and safe so I approve it. Let's give it a couple more days to see if we get another reviewer. If not I will be able to merge it.

But yeah the ldap module isn't very actively maintained and possibly the maintainers just missed your issue. You could eventually propose yourself as a module maintainer (adding your GitHub user to the maintainers value of the module manifest file in a PR will be enough to enable to let you merge and publish improvements in the module).

Thank you for the fix.

sbidoul commented on users_ldap_groups/models/res_company_ldap_operator.py:
```diff
@@ -20,22 +20,33 @@ def operators(self):
         return bool(results)
+
+    def safe_ldap_decode(self, attr):
+        import base64
```
Why not importing at the top?

oh2fih replied:
Good point! As it is imported anyway, and not just if the exception occurs, it doesn't make technical difference to be at the top; moved there.

sbidoul commented on users_ldap_groups/models/res_company_ldap_operator.py:
```diff
@@ -20,22 +20,33 @@ def operators(self):
+        try:
+            return attr.decode()
+        except UnicodeDecodeError:
+            return base64.b64encode(attr)
```
A comment explaining why a base 64 decoding fallback makes sense would be helpful.

oh2fih replied:
A docstring explaining the decision made added for the method.

sbidoul replied:
I'm not familiar with LDAP, but decoding without specifying the encoding and this kind of fallback seem suspicious to me. Does LDAP provide a mechanism to know the encoding of a field in advance, or to know if a field is effectively a binary field?

pedrobaeza replied:
This is encoding, not decoding, so I saw it safe enough, checking also the test.

oh2fih replied:
The `x.decode()` is there in the `_contains()` & `_equals()`, too, and has been part of this module since it was migrated to 12.0 in [[MIG] users_ldap_groups: Migration to 12.0](https://github.com/OCA/server-auth/commit/ad15e3aa96d7d6f04aba840cd1499ad2c3af78ea) ([blame](https://github.com/OCA/server-auth/blame/4501d8a8772e90b99be07ff3a093f4036e6a1a72/users_ldap_groups/models/res_company_ldap_operator.py)). This PR never added it, but only the `base64.b64encode(attr)` as a fallback in case it fails.

Maybe @alexey-pelykh could shed some light on the background behind this decision.

alexey-pelykh replied:
I concur with @oh2fih that such fix should be consistent across the codebase. First and foremost - do we have actual usecase when we need binary data and don't skip the non-UTF8-decodable string?

alexey-pelykh replied:
> So, why not use this safe_ldap_decode() in contains & equals then? Because only the requested LDAP attributes are compared and they should never contain binary data. This is a special case for the query operator for which LDAP returns the entire Active Directory object with all the attributes.

Okay, that explains why only `_query` yet the part "do we actually have any need for binary fields like photos" remain

oh2fih replied:
No, we do not **need** binary fields, but we need to expect to receive them. Another option would be to disregard this data and return an empty value. However, the base64 approach likely makes it easier to debug what data was returned that caused the problem.

alexey-pelykh replied:
I agree on debugging

sbidoul reviewed (commented):
Thanks for submitting this fix.

For next time, note security@odoo-community.org may be contacted for reporting such issues.

sbidoul commented:
This decode-that-encodes-in-base64 looks a bit weird to me. Why is the `decode` necessary in the first place?

Perhaps, to move faster, you could do the security fix in a PR, so I can merge this, and do the decode improvement in another PR and work on that one with folks who are more familiar with this module?

oh2fih commented:
> This decode-that-encodes-in-base64 looks a bit weird to me. Why is the `decode` necessary in the first place?

This simply replaces the original `ldap_entry[1][attr][0].decode()` doing the fix with the least possible modifications to the module. The `Str.decode()` was added in ad15e3aa96d7d6f04aba840cd1499ad2c3af78ea when migrating the module from Odoo 10 to Odoo 12 and seems quite necessary, because it did not work without it on the first attempts to address this bug. The `contains` & `equals` uses the `Str.decode()` (in `x.decode()`), too, so not using it here does not seem coherent, either. Although not explained in the commit, I guess the root cause might be support for non-ASCII names.

So, why not use this `safe_ldap_decode()` in `contains` & `equals` then? Because only the requested LDAP attributes are compared and they should never contain binary data. This is a special case for the `query` operator for which LDAP returns the entire Active Directory object with all the attributes.

Does this satisfactorily explain the situation here?

oh2fih commented:
> Perhaps, to move faster, you could do the security fix in a PR

@sbidoul: There's now another PR https://github.com/OCA/server-auth/pull/659 that only contains the commit fixing the vulnerability. After merging that this PR is only the second commit ahead of upstream `16.0` branch.

oh2fih committed:
[16.0][FIX] users_ldap_groups: safe LDAP decode
The group mapping in query mode fails if LDAP returns binary data in any
of the fields. This adds a function that handles such situation by
base64 encoding it.
The new test test_users_ldap_groups_ldap_returns_binary_data covers
the common case where LDAP return binary data in thumbnailPhoto.

oh2fih commented:
Rebased to fix the checks that didn't pass; the problem was in another module.

Could someone please review & merge this already?

oh2fih commented:
Rebased over `fc906889df964a3f8b50013389dbc585aee0080a` after workflow fixed in https://github.com/OCA/server-auth/issues/693.

oh2fih commented:
@sbidoul It seems difficult to get anyone that actually knows something about LDAP to review this fix. However, this has been running on an instance for months and has fixed a real-life problem there. The problematic LDAP response has been demonstrated in a new test this PR adds, and all the previous tests are passing, too. Could this finally be merged?

pedrobaeza reviewed (approved):
/ocabot merge patch
